internal/messaging/response_handler.go

```diff
@@ -1,147 +1,554 @@
 		}
+		h.logger.Errorf(errMessage)
 		addErrorToResponseHeader(response, errMessage)
 		return true
 	}
-	if !success {
-		addErrorToResponseHeader(response, "buying NFT failed")
-		return true
-	}
 
-	h.logger.Infof("Bought NFT (txID=%s) with ID: %s\n", txID, mintID)
-	response.BuyTransactionId = txID.String()
+	h.logger.Infof("Bought NFT (txID=%s) with ID: %s\n", txID, response.MintTransactionId)
+	response.BuyTransactionId = txID
 	return false
 }
 
-func addErrorToResponseHeader(response *ResponseContent, errMessage string) {
-	response.MintResponse.Header.Status = typesv1alpha.StatusType_STATUS_TYPE_FAILURE
-	response.MintResponse.Header.Alerts = append(response.MintResponse.Header.Alerts, &typesv1alpha.Alert{
-		Message: errMessage,
-		Type:    typesv1alpha.AlertType_ALERT_TYPE_ERROR,
-	})
+// Loads an ABI file
+func loadABI(filePath string) (abi.ABI, error) {
+	file, err := os.ReadFile(filePath)
+	if err != nil {
+		return abi.ABI{}, err
+	}
+	return abi.JSON(strings.NewReader(string(file)))
+}
+
+// Registers a new supplier with the BookingToken contract
+func register(client *ethclient.Client, logger *zap.SugaredLogger, contractABI abi.ABI, privateKey *ecdsa.PrivateKey, supplierName string, bookingTokenAddress common.Address) error {
+	address := crypto.PubkeyToAddress(privateKey.PublicKey)
+	nonce, err := client.PendingNonceAt(context.Background(), address)
+	if err != nil {
+		return err
+	}
+
+	gasPrice, err := client.SuggestGasPrice(context.Background())
+	if err != nil {
+		return err
+	}
+
+	// Check supplier name and set default if empty
+	if supplierName == "" {
+		defaultSupplierName := "Default Supplier"
+		logger.Infof("Supplier name cannot be empty. Setting name to: %s \n", defaultSupplierName)
+		supplierName = defaultSupplierName
+	}
+
+	packed, err := contractABI.Pack("registerSupplier", supplierName)
+	if err != nil {
+		return err
+	}
+
+	gasLimit := uint64(170000)
+
+	tx := types.NewTransaction(nonce, bookingTokenAddress, big.NewInt(0), gasLimit, gasPrice, packed)
+
+	chainID, err := client.NetworkID(context.Background())
+	if err != nil {
+		return err
+	}
+
+	signedTx, err := types.SignTx(tx, types.NewEIP155Signer(chainID), privateKey)
+	if err != nil {
+		return err
+	}
+
+	err = client.SendTransaction(context.Background(), signedTx)
+	if err != nil {
+		return err
+	}
+
+	logger.Infof("Transaction sent!\nTransaction hash: %s\n", signedTx.Hash().Hex())
+
+	// Wait for transaction to be mined
+	receipt, err := waitTransaction(context.Background(), client, logger, signedTx)
+	if err != nil {
+		if gasLimit == receipt.GasUsed {
+			logger.Errorf("Transaction Gas Limit reached. Please use shorter supplier name.\n")
+		}
+		return err
+	}
+
+	return nil
+}
+
+// Mints a BookingToken with the supplier private key and reserves it for the buyer address
+// For testing you can use this uri: "data:application/json;base64,eyJuYW1lIjoiQ2FtaW5vIE1lc3NlbmdlciBCb29raW5nVG9rZW4gVGVzdCJ9Cg=="
+func mint(
```

Comment:
Please, refactor mint and buy to be handler methods. It doesn't make much sense to keep them separated, since they're private, used directly from handler methods and need handler private fields (currently passed as args) to function.